### PR TITLE
Add support for GCP service account credentials 

### PR DESCRIPTION
**Pull request by bensussman**
Add support for GCP service account credentials to be provided in kubernetes configs in addition to the current alternatives (cmd based and google default token source)

/sig gcp

@kubernetes/sig-gcp-pr-reviews

/kind feature

**What this PR does / why we need it**:
This PR adds support for the go-client to authenticate to GCP kubernetes clusters using service account credentials. This is a useful feature because it allows users to configure their kubernetes client authentication in certain contexts that would be difficult to impossible without this, specifically when using "assumed service accounts" (details here https://cloud.google.com/iam/docs/understanding-service-accounts ) in processes that are handling many simultaneous connections to many unique GKE clusters managed by intermediate service accounts that the GKE owners use to give permission to external managers. This came out of a specific need that my organization had when managing our customers' GKE clusters, and is detailed in this SO question: https://stackoverflow.com/questions/55858335/how-to-authenticate-to-multiple-gke-kubernetes-clusters-using-dynamic-and-unique

**Does this PR introduce a user-facing change?**:
It does in-so-far as we now support a new key:value pair inside the `kubeconfig` "service-acct" which expect a value to be a string containing a service account credentials JSON. Example:
```
- name: gke_cluster-name_us-west2-a_project-name
  user:
    auth-provider:
      config:
        service-acct: '{"type": "service_account",  "project_id": "my-proj-24234", ... }'
      name: gcp
```
```release-note
Added support for "service-acct" key for `gcp auth-provider` config.
```

### CODE REVIEW

**Comment by bensussman** [comment]
@fedebongio @yliaog any update on this?

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go
Comment (yliaog) [inline staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: nit: the key string "service-acct" is used twice in the code, better to use a const for it.

**Review by yliaog — APPROVED** [review]

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go
Comment (bensussman) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/gcp/gcp.go]: I pattern matched from the key string `"cmd-path"` which is also used twice and is not a const.

Would you like me to pull both out into constants?

**Comment by yliaog** [comment]
i think it's better to pull both into consts, but it's not something critical, just nit, so i leave it up to you.

**Comment by bensussman** [comment]
Looks like test failures are b/c i changed the input to a function used in test. Attempting to fix now...

**Comment by bensussman** [comment]
@yliaog these test failures don't seem related. Are these known issues with the test, or are these failures because of my change but far downstream?

**Comment by yliaog** [comment]
https://prow.k8s.io/job-history/kubernetes-jenkins/pr-logs/directory/pull-kubernetes-e2e-gce?buildId=1125523999254646786

shows the test history, it has one successful run 1125513062443388928	2019-05-06 21:30:19 +0000 UTC

but it could be just flaky.

**Comment by cjcullen** [comment]
This is really cool, but I worry about:
- Encouraging the use of different kubeconfig entries within a single process.
- Deviating from the existing pattern in all of the Google client libraries.
- Putting long-lived credentials into a file that doesn't have a great programmatic way to rotate them.

Can you give an example of how this AuthProvider would be instantiated for multiple identities in code?

Would it work to construct a Go client directly? E.g.:
```
func ClientFromSAKeyFile(ctx context.Context, filename string, scopes ...string) *rest.RESTClient, error {
	b, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	creds, err := google.CredentialsFromJSON(ctx, b, scopes...)
	if err != nil {
		return nil, err
	}
	rest.RESTClientFor(&rest.Config{
		Transport: &oauth2.Transport{
			Source: creds.TokenSource,
		}
	}), nil
}
```

I think the `--impersonate-service-account` flag is probably the better way to handle this "multiple identities in the same process" problem, but it looked like you ran into some trouble using that.

**Comment by bensussman** [comment]
@cjcullen 

> Encouraging the use of different kubeconfig entries within a single process.

Certainly that's the goal here. I don't really see why you would need a separate process for each remote GKE instance you are managing, especially because Goroutines are so lightweight and most interactions with a GKE instance is I/O bound. For my specific use case, this allows me to dynamically add and remove GKE instances based on customer requests and each is handled by its own service account credentials. Why is this a bad thing?

> Deviating from the existing pattern in all of the Google client libraries.

Not sure what you are referring to here.

> Putting long-lived credentials into a file that doesn't have a great programmatic way to rotate them.

The kube-config is *the* way to authenticate to a kubernetes cluster. The best way to "cut off" credentials is to remove the permissions on the cluster side through RBAC (or however you gave permissions). For our use case, these service-accounts are created for the sole purpose of being granted kube-cluster-permissions.

> I think the `--impersonate-service-account` flag is probably the better way to handle this "multiple identities in the same process" problem, but it looked like you ran into some trouble using that.

It does seem like it could work, but I have been unable to get it to not crash with `ERROR: gcloud crashed (AttributeError): 'unicode' object has no attribute 'utcnow'`. It also requires that the machine that is running the go-code has `gcloud` installed and configured to the IAM user that has permissions for impersonation. This is kind of an onerous requirement given how big the `gcloud` cli is and that we already have a mechanism (and are actively using it) of impersonating from within golang and getting the `service-acct-json` that we need.

This feels like the best way to provide GCP users with a way to authenticate their interactions with their GKE clusters using service-acct-credentials.

**Comment by cjcullen** [comment]
> @cjcullen
> 
> > Encouraging the use of different kubeconfig entries within a single process.
> 
> Certainly that's the goal here. I don't really see why you would need a separate process for each remote GKE instance you are managing, especially because Goroutines are so lightweight and most interactions with a GKE instance is I/O bound. For my specific use case, this allows me to dynamically add and remove GKE instances based on customer requests and each is handled by its own service account credentials. Why is this a bad thing?

My bad, my justification wasn't very clear. I wasn't recommending _not_ managing multiple clusters within a single process. I was just encouraging not using a kubeconfig file to drive it. The kubeconfig file is great when you're writing software that is going to get used by end users in arbitrary ways (like kubectl), and you want to support very flexible configurations (and easy to understand switching between configurations). It isn't an especially good solution for parallel access to multiple clusters using multiple users.

I'm guessing you probably _can_ get a rest.Client built w/ an overriden current_context, but I'd be scared about other places making bad assumptions. 

> 
> > Deviating from the existing pattern in all of the Google client libraries.
> 
> Not sure what you are referring to here.

All of the Google Cloud client libraries follow the same structure and order for obtaining credentials:
1. Try the Google App Engine credential service
2. See if `GOOGLE_APPLICATION_CREDENTIALS` is set.
3. Check the "well known" credential file managed by gcloud.
4. Try the Google Compute Engine token endpoint.

I wouldn't want somebody to use this new mechanism, and then be confused that it doesn't match the behavior of other Google-authenticating clients.

But, I don't think application portability is necessarily your goal, so you can maybe ignore this point.

> 
> > Putting long-lived credentials into a file that doesn't have a great programmatic way to rotate them.
> 
> The kube-config is _the_ way to authenticate to a kubernetes cluster. The best way to "cut off" credentials is to remove the permissions on the cluster side through RBAC (or however you gave permissions). For our use case, these service-accounts are created for the sole purpose of being granted kube-cluster-permissions.

For your specific case, removing RBAC permissions is probably enough to deal with a credential compromise. In the general case, you can't guarantee that this identity hasn't been granted permissions elsewhere, and you can't guarantee that this key is the only credential being used by this identity. So, revoking RBAC permissions might not be enough, or it might be too heavy.

> 
> > I think the `--impersonate-service-account` flag is probably the better way to handle this "multiple identities in the same process" problem, but it looked like you ran into some trouble using that.
> 
> It does seem like it could work, but I have been unable to get it to not crash with `ERROR: gcloud crashed (AttributeError): 'unicode' object has no attribute 'utcnow'`. It also requires that the machine that is running the go-code has `gcloud` installed and configured to the IAM user that has permissions for impersonation. This is kind of an onerous requirement given how big the `gcloud` cli is and that we already have a mechanism (and are actively using it) of impersonating from within golang and getting the `service-acct-json` that we need.

Yeah, I agree with that. Dependence on another binary, and shelling-out in general feels kinda gross for the in-process case.

We (Google) have been a little too quick to add Google-specific stuff to kubernetes/kubernetes in the past, and we're trying to move away from it. I doubt we'd even be able to get the original "gcp" authProvider in today.

> 
> This feels like the best way to provide GCP users with a way to authenticate their interactions with their GKE clusters using service-acct-credentials.

I really want to find a way to solve your problem, because I'm excited that you are using GKE, and especially happy that you went out and wrote the code you needed and sent out a PR.

It sounds like this change solves your use case well, but I don't think it is a good approach in the general case for two reasons:
1. It encourages poor credential hygiene.
2. It adds more Google-specific code to kubernetes/kubernetes.

I want to solve your problem though, so let's keep talking about it.

**Comment by bensussman** [comment]
Thanks @cjcullen 

Is there documentation on how to construct a `kubernetes clientset` such that it uses your own `RESTClient`, but also is correctly configured with the appropriate values? For context, this is how we are creating kubernetes clients in our go-code:

```
clientConfig, err := clientcmd.NewClientConfigFromBytes([]byte(config))
if err != nil {
	return nil, errors.Wrap(err)
}
restClientConfig, err := clientConfig.ClientConfig()
if err != nil {
	return nil, errors.Wrap(err)
}
clientset, err := kubernetes.NewForConfig(restClientConfig)
if err != nil {
	return nil, errors.Wrap(err)
}
```

It doesn't seem like the "right way" to construct a kubernetes client is by manually patching together the different values inside the different nested structs, but if that's what you're saying I guess that's the agreed on best practice? It would be helpful to see some documentation or example code that parses a kubeconfig and then generates a `clientset` that uses an auth provider that is powered by service-account-json.

Our current solution works by reading a kubeconfig and updating the "auth" section to string-format-in the service-account-json such that the kubernetes golang auth code edited in this PR will get the right tokens and then call `NewClientConfigFromBytes` with that kubeconfig. It's working for us internally (i simply register a new auth provider `gcp-svc` temporarily while this PR remains unmerged) but it's not the most elegant. However if i understand your proposal fully, it seems even less elegant, we need to:

1. Parse the kubeconfig with a YAML parser and extract the values we need
2. Directly construct the `ClientConfig` struct using the values extracted from 1
3. Create a RESTClient that uses the GCP token source powered by the service-account-json generated from an API request to GCP
4. Inject/clobber the RESTClient into the `kubernetes clientset`

Do you have a more elegant solution? Or is this seemingly less elegant solution proposed above better for some reason? How would you like to keep talking about it? We could schedule a call? Let me know!

**Comment by ahmetb** [comment]
I'm not sure why this PR is necessary.

You can just set `GOOGLE_APPLICATION_CREDENTIALS` environment variable to `/path/to/key.json` and clear the access token in the kubeconfig context data, and the `kubectl` will automatically use the service account credentials you passed to authenticate.

**Comment by bensussman** [comment]
An environment variable is global @ahmetb and not specific to each individual kubernetes cluster you are authenticating with. In addition maintaining a bunch of different files on disk for each `key.json` is significantly more onerous than the solution created by this PR.

**Comment by mikedanese** [comment]
Why do you need a different service account per cluster?

**Comment by bensussman** [comment]
@mikedanese I can think of lots of use cases:

The use case that we are using this for at my company ( https://spell.run ) is to manage our customers' clusters. This allows customers to keep their private data and traffic within their own VPC, while not having to worry about the administration of a kubernetes cluster.

Another use case is if you have different accounts with different levels of privileges that you want a single system to administer (like your CI for example)

Another use case is if you frequently rotate service account credentials, or have an external service that provides credentials, and want to control the credentials at runtime.

**Comment by cjcullen** [comment]
Sorry for dropping off on this...

> It doesn't seem like the "right way" to construct a kubernetes client is by manually patching together the different values inside the different nested structs

I'm more comfortable in this sort of situation knowing which pieces I've configured. But I guess there's a tradeoff for convenience using an already defined file format.

> 1. Parse the kubeconfig with a YAML parser and extract the values we need
> 2. Directly construct the ClientConfig struct using the values extracted from 1
> 3. Create a RESTClient that uses the GCP token source powered by the service-account-json generated from an API request to GCP
> 4. Inject/clobber the RESTClient into the kubernetes clientset

I think if you went this route, the only clobbering you'd do is the Transport in the rest.Config, like in my code snippet above.

> This allows customers to keep their private data and traffic within their own VPC, while not having to worry about the administration of a kubernetes cluster.

Using different plaintext private keys accessible within the same process isn't giving you any more of a boundary between clusters, and it just expands your credential management/rotation problem.

Alternatively, you could have a single identity for your management control plane, and grant that identity the serviceAccountTokenCreator role on the per-cluster service accounts, if you want a more capabilities-like authZ model. This way, you still get the benefit of using multiple service accounts (it likely takes two code bugs to do a thing to the wrong cluster), but you don't have so many private keys floating around. I think this is what @mikedanese was likely getting at.

If you end up deciding the immediate convenience of "everything in kubeconfig" wins, I'd still rather the key material not be included in the kubeconfig file. If this were modeled as `service-acct-file` a la the `client-key` field in [clientcmdapi.AuthInfo](https://github.com/kubernetes/client-go/blob/master/tools/clientcmd/api/types.go#L96), it would at least solve my "hard to programmatically rotate" gripe (still not _great_ credential hygiene, but better). I'm still not thrilled about adding more GCP specific stuff (because we end up implicitly on the hook to support it), but if there's obviously no better way, I think we can make a decent case.

**Comment by mikedanese** [comment]
So the reason I really don't like this change is that it promotes the downloading of keys which is something that we really don't want people to do. Instead can you configure the client with a kubeconfig like this:

```yaml
- name: gke_mikedanese-k8s_us-central1-c_foobar
  user:
    auth-provider:
      config:
        cmd-args: auth print-access-token --format=json --impersonate-service-account=customeraccount@myproj.iam.gserviceaccount.com
        cmd-path: /home/mikedanese/google-cloud-sdk/bin/gcloud
        expiry-key: '{.token_expiry}'
        token-key: '{.access_token}'
      name: gcp
```

And grant the gcloud user permission token creator:

```console
gcloud iam service-accounts add-iam-policy-binding \
  --role roles/iam.serviceAccountTokenCreator \
  --member "user:mikedanese@example.com" \
  customeraccount@myproj.iam.gserviceaccount.com
```

Does this look like it would solve you use case.

Note: we fixed a bug in gcloud last week to allow this so you might have to run `gcloud components update` for this to work.

**Comment by bensussman** [comment]
No @mikedanese  this does not fix our use case. Requiring `gcloud` to be installed on our golang API servers does not feel like a good solution.

Since there is so much pushback from the kube maintainers, I will be closing this PR. One change we have made internally is to stop using `serviceAccountKeyAdmin` to generate service-account-JSON and use `serviceAccountTokenCreator` instead  to generate tokens directly. This means that we can build our own custom `TokenSource` which uses `iamcredentials.GenerateAccessTokenRequest` to create a token.

This STILL REQUIRES a custom `RegisterAuthProviderPlugin()` so that we can leverage the custom `TokenSource` from above. I would love to NOT use a kubeconfig, could you provide an example or a link to an example @cjcullen where you construct a `*kubernetes.Clientset` given a `server IP Address`, `certificate-authority-data`, and a custom `*rest.RESTClient`? The methods i see are `NewDefaultClientConfig`, `NewNonInteractiveClientConfig`, and `NewInteractiveClientConfig`, but none of these take the inputs that I have. Given the sparse documentation, you would have a more compelling case about what the "right way" to construct a `*kubernetes.Clientset` is.

**Comment by ahmetb** [comment]
> Requiring `gcloud` to be installed on our golang API servers does not feel like a good solution.

You can avoid installing gcloud by clearing the gcloud parts of kubeconfig (just keep `auth:{name: gcp}` and setting the service account .json key file's path in `GOOGLE_APPLICATION_CREDENTIALS` environment variable (unless I'm mistaken).

**Comment by bensussman** [comment]
@ahmetb that does not solve our problem, because `GOOGLE_APPLICATION_CREDENTIALS` is global and we need to have different credentials per cluster. Please read the conversation above for more details.

**Comment by lnhrdt** [comment]
Similar to @bensussman, I am trying to manage multiple GKE clusters using with different service accounts for each cluster (the clusters are owned by different organizations). I am working in an environment where we've been trying to avoid adding the `gcloud` CLI to the mix and instead continue relying on the more familiar `kubectl` whenever possible as a way to keep our system simple and generic.

So far we've opted to manually change the `GOOGLE_APPLICATION_CREDENTIALS` environment variable to correspond with the cluster set in the kubeconfig context. However it would further simplify our system if the GCP service account JSON could be set directly in the kubeconfig, rather than relying on the global `GOOGLE_APPLICATION_CREDENTIALS` environment variable. I would expect the JSON could be configured as either a path to the file or embedded, just like certificate authority certificates etc.

@bensussman since your PR wasn't merged in, have you since developed any other approaches? Besides building or modifying our own client as you've done.

**Comment by bensussman** [comment]
@lnhrdt We have this solved at my organization using the solution in this PR, copy-pasted into a file that registers itself as a new auth provider named `gcp-svc`. This PR was designed to share our changes with others as this solves our problem quite well, however given that it didn't get merged, you can do the same as us: copy paste this file, and update the "registration" code to have a new unique name:

```
func init() {
	if err := restclient.RegisterAuthProviderPlugin("gcp-svc", newGCPAuthProvider); err != nil {
		klog.Fatalf("Failed to register gcp auth plugin: %v", err)
	}
}
```

Then use `gcp-svc` as your auth provider instead of `gcp` and it will use *your* registered provider, which has the additional service account credential parsing contained in this PR.

We've been using it for >18 months now!

**Comment by arun11299** [comment]
I really wanted this feature in. Seems like GCP team wants people to use gcloud as much as possible.

**Comment by arun11299** [comment]
@bensussman  Any reason why you did not go for a plain kubernetes service account based kubeconfig ? I am not sure myself but are there any reasons it would not have worked for your usecase ?

**Comment by bensussman** [comment]
@arun11299 I'm not sure what you mean. If you read the PR description carefully, I believe it outlines specifically the use case that this PR enables:

> This PR adds support for the go-client to authenticate to GCP kubernetes clusters using service account credentials. This is a useful feature because it allows users to configure their kubernetes client authentication in certain contexts that would be difficult to impossible without this, specifically when using "assumed service accounts" (details here https://cloud.google.com/iam/docs/understanding-service-accounts ) in processes that are handling many simultaneous connections to many unique GKE clusters managed by intermediate service accounts that the GKE owners use to give permission to external managers

Fundamentally, this PR is important in the aforementioned case: when you have MULTIPLE SIMULTANEOUS connections to many unique GKE clusters managed by intermediate service accounts that the GKE owners have granted permission to for external management.

**Comment by arun11299** [comment]
Thanks for the clarification. I understand that your usecase is to support authentication using the service key file. We also needed something like that, but something which would work for all both python and Go clients.
I was referring to a solution where the GKE owner would directly provide you with a kubeconfig file having token of a service account with required roles. See https://medium.com/google-cloud/authing-w-kubernetes-engine-service-accounts-ae752b46ed18
This is an alternate solution if you or anybody else visiting here would be interested.

**Comment by bensussman** [comment]
Thanks @arun11299 ! I think the problem with this is that you are using token auth to the K8S cluster, which is generally not advised! We prefer to use Service Account authentication as it's easier to track and revoke and manage from a normal GCP IAM workflow.

**Pull request closed without merging**
